src/main/java/org/opensearch/agent/tools/utils/clustering/ClusteringHelper.java

```diff
@@ -0,0 +1,525 @@
+
+                List<String> clusterCentroids = processCluster(kMeansCluster, vectors, indexTraceIdMap, clusterIdx);
+                finalCentroids.addAll(clusterCentroids);
+            }
+
+        } catch (Exception e) {
+            log.warn("K-means clustering failed, falling back to hierarchical clustering only: {}", e.getMessage());
+            // Fallback to hierarchical clustering only
+            finalCentroids = performHierarchicalClustering(vectors, indexTraceIdMap);
+        }
+
+        return finalCentroids;
+    }
+
+    /**
+     * Processes a single K-means cluster
+     *
+     * @param kMeansCluster List of indices in the K-means cluster
+     * @param vectors Original vector array
+     * @param indexTraceIdMap Original mapping from indices to trace IDs
+     * @param clusterIdx Index of the cluster (for logging)
+     * @return List of trace IDs representing cluster centroids
+     */
+    private List<String> processCluster(
+        List<Integer> kMeansCluster,
+        double[][] vectors,
+        Map<Integer, String> indexTraceIdMap,
+        int clusterIdx
+    ) {
+        if (kMeansCluster.isEmpty()) {
+            return List.of();
+        }
+
+        if (kMeansCluster.size() == 1) {
+            return List.of(indexTraceIdMap.get(kMeansCluster.getFirst()));
+        }
+
+        if (kMeansCluster.size() > 500) {
+            log.debug("The cluster size is greater than 500, performing partitioned clustering");
+            return performHierarchicalClusteringOfPartition(kMeansCluster, vectors, indexTraceIdMap);
+        }
+
+        log.debug("Applying hierarchical clustering to K-means cluster {} with {} points", clusterIdx, kMeansCluster.size());
+
+        // Extract vectors for this K-means cluster
+        double[][] clusterVectors = extractVectors(kMeansCluster, vectors);
+        Map<Integer, String> clusterIndexTraceIdMap = createTraceIdMapping(kMeansCluster, indexTraceIdMap);
+
+        // Apply hierarchical clustering within this K-means cluster
+        return performHierarchicalClustering(clusterVectors, clusterIndexTraceIdMap);
+    }
+
+    /**
+     * Perform K-means clustering using Apache Commons Math3
+     *
+     * @param vectors Input vectors for clustering
+     * @param numClusters Number of K-means clusters
+     * @return List of clusters, each containing indices of points in that cluster
+     * @throws RuntimeException if clustering fails
+     */
+    private List<List<Integer>> performKMeansClustering(double[][] vectors, int numClusters) {
+        if (vectors == null || vectors.length == 0) {
+            return new ArrayList<>();
+        }
+
+        if (numClusters <= 0) {
+            numClusters = 1;
+        }
+
+        // Cap number of clusters to vector size
+        numClusters = Math.min(numClusters, vectors.length);
+
+        try {
+            KMeansPlusPlusClusterer<DoublePoint> clusterer = createKMeansClusterer(numClusters);
```

Comment:
Will we have memory concern here? Kmeans is a memory-saved method but it's better to have some test or have a memory detection like https://github.com/opensearch-project/ml-commons/blob/main/plugin/src/main/java/org/opensearch/ml/breaker/MemoryCircuitBreaker.java

Reply:
@PauiC will help to paster the benchmark result here for memory and cpu usage

Reply:
According to the benchmark, the memory usage of kmeans looks safe.